docs: generated JSON schema reference, tidy up README and contrib docs

coopernetes opened the pull request:
- shorten the README to include minimal instructions. Link to doc site
  for detailed documentation.
- remove boilerplate from README for contributing steps. Moved those
  guidelines to CONTRIBUTING doc
- use simplified npx commands for executing git-proxy
- using github.com/coveooss/json-schema-for-humans to generate a
  readable document in Markdown form to display in doc site
- include a config used by json-schema-for-humans (jhsf.config.json)
  so that doc can be generated repeatably
- add npm script to execute the CLI with the included config and format
  into a proper MDX doc for Docusaurus to serve

Testing npx with the new release:

```bash
$ npx -- @finos/git-proxy@1.1.0
Need to install the following packages:
  @finos/git-proxy@1.1.0
Ok to proceed? (y) 
npm WARN deprecated mkdirp@0.3.0: Legacy versions of mkdirp are no longer supported. Please update to mkdirp 1.x. (Note that the API surface has changed to use Promises in 1.x.)
npm WARN deprecated @material-ui/styles@4.11.5: Material UI v4 doesn't receive active development since September 2021. See the guide https://mui.com/material-ui/migration/migration-v4/ to upgrade to v5.
npm WARN deprecated @material-ui/core@4.12.4: Material UI v4 doesn't receive active development since September 2021. See the guide https://mui.com/material-ui/migration/migration-v4/ to upgrade to v5.
creating user
        user=admin,
        gitAccount=none
        email=admin@place.com,
        admin=true
Service Listening on 8080
Listening on 8000

$ npx -- @finos/git-proxy --help
Usage: git-proxy [options]

Options:
      --help      Show help                                            [boolean]
      --version   Show version number                                  [boolean]
  -v, --validate  Check the proxy.config.json file in the current working
                  directory for validation errors.
  -c, --config    Path to custom git-proxy configuration file.
                                                  [default: "proxy.config.json"]
$ npx -- @finos/git-proxy --version
1.1.0
$ npx -- @finos/git-proxy
Service Listening on 8080
Listening on 8000
```

Running the new script:
```
$ npm run gen-schema-doc

> @finos/git-proxy@1.1.0 gen-schema-doc
> node ./scripts/doc-schema.js

== Generating schema.md ==
== Generated schema.md in 0:00:00.057090 ==

Wrote schema reference to ./website/docs/configuration/reference.mdx
```

coopernetes commented:
@maoo @JamieSlome I'm assuming the netlify preview won't deploy until #325 is merged?

JamieSlome commented:
@coopernetes - these deployments are being weird. I'm getting some PRs that are deploying, some that aren't.

maoo commented:
I manually re-triggered the Netlify build, preview is available on https://deploy-preview-327--endearing-brigadeiros-63f9d0.netlify.app/ and looks great, TY @coopernetes !

Also left a comment on #325 , right now netlify behaves a bit randomly and would be great to solve it.

JamieSlome commented on README.md:
````diff
@@ -46,234 +46,59 @@ Git Proxy deploys custom push protections and policies on top of Git. It is a hi
 ```bash
-npm install -g @finos/git-proxy@1.1.0
+$ git clone https://github.com/octocat/Hello-World.git && cd Hello-World
+$ gh repo fork
````
@coopernetes - okay assuming that folks will know what this is?

coopernetes replied:
Trying my best here to give a shortest possible introduction to git-proxy. Most of the minutiae is captured in the doc site's quick start guide so didn't want to reinvent here and would rather direct users to that guide which is more end-to-end and comprehensive. Alternatively, we can have just `Install & run git-proxy` section above and omit the push example. Something like:

````markdown
## Getting Started
Install & run git-proxy:

```bash
$ npx -- @finos/git-proxy
Need to install the following packages:
  @finos/git-proxy@1.1.0
Ok to proceed? (y) 
...
creating user
        user=admin,
        gitAccount=none
        email=admin@place.com,
        admin=true
Service Listening on 8080
Listening on 8000
```
You can now push commits directly through the proxy server and start implementing your own organization's controls. For a more comprehensive demo, please see the [Quick start guide](https://git-proxy.finos.org/docs/category/quickstart) on the project's documentation.
````

JamieSlome replied:
@coopernetes - I like the length and simplicity of your introduction, we should keep this. I'm more referring to the use of the `gh` CLI in the steps?

coopernetes replied:
Good point. I will leave a small bash-style comment.

JamieSlome commented on README.md:
```diff
@@ -46,234 +46,59 @@ Git Proxy deploys custom push protections and policies on top of Git. It is a hi
+Compressing objects: 100% (2/2), done.
+Writing objects: 100% (3/3), 946 bytes | 946.00 KiB/s, done.
+Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
+remote: ERR     Rejecting repo yourGithubUser/Hello-World.git not in the authorisedList
```
Does it make sense to show a failed attempt in the `README.md` instead of an intercepted one?

coopernetes replied:
See https://github.com/finos/git-proxy/pull/327#discussion_r1375265790 - I'm happy to remove this completely since the user journey is captured in the quick start guide.

JamieSlome replied:
@coopernetes - just scrolled down to this comment. I'd agree, let's just show running Git Proxy and point the developer to the docs 👍

coopernetes replied:
Did you want me to remove the failing push output from the README? I already included a note describing that you should expect a push to fail without modifying the default configuration, which I believe is important to state right up front. A more complete, end-to-end example is in documentation and makes sense to direct new users there.

JamieSlome replied:
Should we just remove it entirely, as you suggested earlier and use more of the estate to point the viewer to the documentation site instead? @coopernetes

JamieSlome reviewed (commented):
Couple a of questions there for ya @coopernetes. Beautiful PR though ❤️

JamieSlome reviewed (commented):
Updated some of my comments; once addressed, this PR is ready to go 👍

JamieSlome commented on website/docs/usage.mdx:
```diff
@@ -12,9 +12,9 @@ git-proxy
 ### Using [npx instead of npm](https://www.freecodecamp.org/news/npm-vs-npx-whats-the-difference/)
 
-To directly run Git Proxy without any installation, run:
+You can install & run git-proxy in one step using npx:
```
```suggestion
You can install & run `git-proxy` in one step using npx:
```

JamieSlome commented on website/docs/configuration/reference.mdx:
```diff
@@ -0,0 +1,311 @@
+description: JSON schema reference documentation for Git Proxy
+---
+
+# Git Proxy configuration file
```
```suggestion
# Schema Reference
```

JamieSlome commented on website/docs/configuration/overview.mdx:
```diff
@@ -0,0 +1,64 @@
+---
+
+### Default configuration
+
+On installation, Git Proxy ships with an out-of-the-box configuration. This is fine for
+demonstration purposes but is likely not what you want to deploy into your environment.
+
+The default configuration can be found here: [default proxy.config.json](https://github.com/finos/git-proxy/blob/main/proxy.config.json)
+
```
```suggestion

On installation, Git Proxy ships with an [out-of-the-box configuration](https://github.com/finos/git-proxy/blob/main/proxy.config.json). This is fine for
demonstration purposes but is likely not what you want to deploy into your environment.
```

coopernetes commented:
@JamieSlome all the review comments should now be resolved and we have a lean README with minimal instructions. Leaving the full e2e working example to the docs. Let me know if any further changes are required.

JamieSlome commented on README.md:
````diff
@@ -46,234 +46,44 @@ Git Proxy deploys custom push protections and policies on top of Git. It is a hi
 ```bash
-npm install -g @finos/git-proxy@1.1.0
+$ git clone https://github.com/octocat/Hello-World.git && cd Hello-World
+# The below command is using the GitHub official CLI to fork the repo that is cloned.
+# You can also fork on the GitHub UI. For usage details on the CLI, see https://github.com/cli/cli
+# You must fork the repo into our own account to demonstrate pushing through Git Proxy since it
+# is necessary to have write access on the repo for git-proxy to function. If you do not have push
+# access, you will receive a 403 error.
````
Can we condense this for readability?

JamieSlome commented on README.md:
````diff
@@ -46,234 +46,41 @@ Git Proxy deploys custom push protections and policies on top of Git. It is a hi
 ```
 
-## Run Git Proxy ⚡
+Using the default configuration, Git Proxy intercepts the push and _blocks_ it because the forked repository 
+is not listed in the `authorisedList` of repos allowed through the proxy. To allow pushes to your fork, you must configure Git Proxy.
````
```suggestion
Using the default configuration, Git Proxy intercepts the push and _blocks_ it. To enable code pushing to your fork via Git Proxy, add your repository URL into the Git Proxy config file (`proxy.config.json`). For more information, refer to [Documentation](#Documentation).
```

JamieSlome commented on README.md:
````diff
@@ -46,234 +46,40 @@ Git Proxy deploys custom push protections and policies on top of Git. It is a hi
 ```
 
-## Run Git Proxy ⚡
+Using the default configuration, Git Proxy intercepts the push and _blocks_ it. To enable code pushing to your fork via Git Proxy, add your repository URL into the Git Proxy config file (`proxy.config.json`). For more information, refer to [Documentation](#Documentation).
````
```suggestion
Using the default configuration, Git Proxy intercepts the push and _blocks_ it. To enable code pushing to your fork via Git Proxy, add your repository URL into the Git Proxy config file (`proxy.config.json`). For more information, refer to [Documentation](##Documentation).
```

coopernetes commented:
Hmm... @maoo know why this is failing? I swear, this was deploying fine on a previous commit. And I can see in the MDX doc that we have matching closing tags so I don't know why the doc build is failing.

https://app.netlify.com/sites/endearing-brigadeiros-63f9d0/deploys/655f767daa9c240008ef812e
```
10:58:28 AM: [ERROR] Client bundle compiled with errors therefore further build is impossible.
10:58:28 AM: Error: MDX compilation failed for file "/opt/build/repo/website/docs/configuration/reference.mdx"
10:58:28 AM: Cause: Expected a closing tag for `</strong>` (19:1-19:10) before the end of `paragraph`
```
I can try rendering a different template using the new generator tooling for the JSON schema.

coopernetes commented:
The build error is related to compliance with MDX that the json-schema-for-humans tool doesn't account for at the moment in its plain Markdown template. Related: https://github.com/coveooss/json-schema-for-humans/issues/223

I'll push up a fix in the schema doc generation script to account for this.

coopernetes commented:
@JamieSlome @maoo the new docs are now fixed and rebased with all the updates from main. The site is built successfully with the generated schema reference doc. Preview is available here:

https://deploy-preview-327--endearing-brigadeiros-63f9d0.netlify.app/docs/configuration/reference

maoo reviewed (approved):
LGTM!

JamieSlome reviewed (approved):
LGTM! 🍰
